crates/turbopack-ecmascript/src/tree_shake/mod.rs

```diff
@@ -0,0 +1,719 @@
+                comments: comments.clone(),
+                source_map: source_map.clone(),
+                eval_context,
+            }))
+        }
+        ParseResult::Unparseable => Ok(ParseResultVc::cell(ParseResult::Unparseable)),
+        ParseResult::NotFound => Ok(ParseResultVc::cell(ParseResult::NotFound)),
+    }
+}
+
+impl EcmascriptModulePartAssetVc {
+    pub async fn from_splitted(
+        module: EcmascriptModuleAssetVc,
+        part: ModulePartVc,
+    ) -> Result<Self> {
+        let split_data = split(module.origin_path(), module.parse());
+        let result = split_data.await?;
+        let part = part.await?;
+
+        let key = match &*part {
+            ModulePart::ModuleEvaluation => Key::ModuleEvaluation,
+            ModulePart::Export(export) => Key::Export(export.await?.to_string()),
+        };
+
+        let chunk_id = match result.data.get(&key) {
+            Some(id) => *id,
+            None => return Err(anyhow!("could not find chunk id for module part {:?}", key)),
+        };
+
+        Ok(EcmascriptModulePartAsset {
+            full_module: module,
+            chunk_id,
+            split_data,
+        }
+        .cell())
+    }
+}
+
+#[turbo_tasks::value_impl]
+impl Asset for EcmascriptModulePartAsset {
+    #[turbo_tasks::function]
+    fn content(&self) -> AssetContentVc {
+        todo!()
+    }
+
+    #[turbo_tasks::function]
+    fn references(&self) -> AssetReferencesVc {
+        todo!()
+    }
+
+    #[turbo_tasks::function]
+    async fn ident(&self) -> Result<AssetIdentVc> {
+        let inner = self.full_module.ident().await?;
+
+        let query = StringVc::cell(format!("part={}", self.chunk_id).into());
+        Ok(AssetIdentVc::new(Value::new(AssetIdent {
+            query: Some(query),
+            ..(*inner).clone()
+        })))
+    }
+}
+
+#[turbo_tasks::value_impl]
+impl EcmascriptChunkPlaceable for EcmascriptModulePartAsset {
+    #[turbo_tasks::function]
+    async fn as_chunk_item(
+        self_vc: EcmascriptModulePartAssetVc,
+        context: ChunkingContextVc,
+    ) -> Result<EcmascriptChunkItemVc> {
+        let s = self_vc.await?;
+
+        Ok(EcmascriptModulePartChunkItem {
+            module: self_vc,
+            context,
+            chunk_id: s.chunk_id,
+            full_module: s.full_module,
+            split_data: s.split_data,
+        }
+        .cell()
+        .into())
+    }
+
+    #[turbo_tasks::function]
+    async fn get_exports(self_vc: EcmascriptModuleAssetVc) -> Result<EcmascriptExportsVc> {
+        Ok(self_vc.analyze().await?.exports)
+    }
+}
+
+#[turbo_tasks::value_impl]
+impl ChunkableAsset for EcmascriptModulePartAsset {
+    #[turbo_tasks::function]
+    async fn as_chunk(self_vc: EcmascriptModulePartAssetVc, context: ChunkingContextVc) -> ChunkVc {
+        EcmascriptChunkVc::new(context, self_vc.as_ecmascript_chunk_placeable()).into()
+    }
+}
+
+#[turbo_tasks::value]
+pub struct EcmascriptModulePartChunkItem {
```

Comment:
Could we move the asset and the chunk item into separate files? `asset.rs` and `chunk_item.rs`